nightingale_manipulation/src/nightingale_manipulation/manipulation_action_client.py

```diff
@@ -2,29 +2,97 @@
 
 import actionlib
 import rospy
+import math
 
 # joint
 from moveit_action_handlers.msg import PropertyValuePair
 from moveit_action_handlers.msg import MoveToJointsMoveItAction
 from moveit_action_handlers.msg import MoveToJointsMoveItGoal
+from moveit_action_handlers.msg import MoveToJointsMoveItResult
 
+from moveit_action_handlers.msg import ActionResultStatusConstants
 from sensor_msgs.msg import JointState
 
 # cartesian
 from moveit_action_handlers.msg import MoveToPoseMoveItAction
 from moveit_action_handlers.msg import MoveToPoseMoveItGoal
+from moveit_action_handlers.msg import MoveToPoseMoveItResult
 from moveit_action_handlers.msg import PoseStamped
-from geometry_msgs.msg import Pose
 
 # forward kinematics
-# from urdf_parser_py.urdf import URDF
-# from pykdl_utils.kdl_kinematics import KDLKinematics
-
+from urdf_parser_py.urdf import Robot
+import tf_conversions, tf2_ros
+from geometry_msgs.msg import Pose as GeometryPose
+from geometry_msgs.msg import Quaternion, Point
 
 # TODO: replace with service call, requires refactoring of service to include gripper info
 CFG = rospy.get_param("/nightingale_utils/joint_configurations")
 
 
+# overriding the geometry msgs Pose class to add tolerance on equivalence check
+class Pose(GeometryPose):
+    def __init__(self, tolerance=0.01):
+        super.__init__()
+        self.tol = tolerance
+
+    def __eq__(self, other):
+        if type(other) not in [GeometryPose, Pose]:
+            return False
+        return all(
+            [
+                math.isclose(self.position.x, other.position.x, rel_tol=self.tol),
+                math.isclose(self.position.y, other.position.y, rel_tol=self.tol),
+                math.isclose(self.position.z, other.position.z, rel_tol=self.tol),
+                math.isclose(self.orientation.x, other.orientation.x, rel_tol=self.tol),
+                math.isclose(self.orientation.y, other.orientation.y, rel_tol=self.tol),
+                math.isclose(self.orientation.z, other.orientation.z, rel_tol=self.tol),
+                math.isclose(self.orientation.w, other.orientation.w, rel_tol=self.tol),
+            ]
+        )
+
+
+class Orientation:
+    def __init__(self, roll=None, pitch=None, yaw=None, tolerance=0.01):
+        self.roll = roll
+        self.pitch = pitch
+        self.yaw = yaw
+        self.tol = tolerance
+
+    def __eq__(self, other):
+        if not type(other) is Orientation:
+            return False
+        return all(
+            [
+                math.isclose(self.roll, other.roll, rel_tol=self.tol),
+                math.isclose(self.pitch, other.pitch, rel_tol=self.tol),
+                math.isclose(self.yaw, other.yaw, rel_tol=self.tol),
+            ]
+        )
+
+
+def quaternion_to_euler(q: Quaternion):
```

Comment:
There is a function in tf_conversions to do this already

Reply:
I spent so long looking for an existing function -_-